## **Linear Algebra - Dr. Malek**
Amir Hossein Mansoori - Student No. 92943069 - Computer Assignment #02

## Part 1

First, we import the required modules:

In [ ]:
import numpy as np

The we compute different powers of $A$ (from 1 to 10):

In [ ]:
a = np.array([[0, 1], [1, 1]])
n = 10

a_power = [np.linalg.matrix_power(a, i) for i in range(1, n+1)]

print(*a_power, sep='\n\n')

[[0 1]
 [1 1]]

[[1 1]
 [1 2]]

[[1 2]
 [2 3]]

[[2 3]
 [3 5]]

[[3 5]
 [5 8]]

[[ 5  8]
 [ 8 13]]

[[ 8 13]
 [13 21]]

[[13 21]
 [21 34]]

[[21 34]
 [34 55]]

[[34 55]
 [55 89]]


It appears that
$
A^m = \begin{bmatrix}
Fib(m-1) & Fib(m) \\
Fib(m) & Fib(m + 1)
\end{bmatrix}
$,
where $Fib(n) = Fib(n - 1) + Fib(n - 2), Fib(0) = 0, Fib(1) = 1$.

## Part 2

First, we compute $A^{10}$ directly:

In [ ]:
a_10 = np.linalg.matrix_power(a, 10)

print('A^10 = ', a_10)

A^10 =  [[34 55]
 [55 89]]


Now to calculate $A^{10}$ using eigenvalues: We start by calulating eigenvalues from given eigenvectors:

$Av_1 = 
\begin{bmatrix}
0 & 1 \\
1 & 1
\end{bmatrix}
\begin{bmatrix}
2 \\
1 + \sqrt{5}
\end{bmatrix}
=
\lambda_1 v_1
$

In [ ]:
v_1 = np.array([[2], [1 + (5 ** 0.5)]])
a_v1 = np.matmul(a, v_1)

print('A * v1 = ', a_v1)

lambda_1 = a_v1[0][0] / v_1[0][0]

print('lambda_1 = ', lambda_1)

A * v1 =  [[3.23606798]
 [5.23606798]]
lambda_1 =  1.618033988749895


$
\Rightarrow \lambda_1 \approx 1.618
$

Now for $\lambda_2$:

In [ ]:
v_2 = np.array([[2], [1 - (5 ** 0.5)]])
a_v2 = np.matmul(a, v_2)

print('A * v2 = ', a_v2)

lambda_2 = a_v2[0][0] / v_2[0][0]

print('lambda_2 = ', lambda_2)

A * v2 =  [[-1.23606798]
 [ 0.76393202]]
lambda_2 =  -0.6180339887498949


$
\Rightarrow \lambda_2 \approx -0.618
$

Now we form the diagnolized matrix:

$A = PDP^{-1} \Rightarrow A^{10} = PD^{10}P^{-1}
=
\begin{bmatrix}
2 & 2 \\
1 + \sqrt{5} & 1 - \sqrt{5}
\end{bmatrix}
\begin{bmatrix}
\lambda_1 & 0 \\
0 & \lambda_2
\end{bmatrix}^{10}
\begin{bmatrix}
2 & 2 \\
1 + \sqrt{5} & 1 - \sqrt{5}
\end{bmatrix}^{-1}
$

In [ ]:
p = np.column_stack([v_1, v_2])
print('P = ', p)
print()

d = np.array([[lambda_1, 0], [0, lambda_2]])
d_10 = d ** 10 # memberwise power

p_inverse = np.linalg.inv(p)

a_10_diagnolized = np.matmul(np.matmul(p, d_10), p_inverse)

print('A^10 = ', a_10_diagnolized)

P =  [[ 2.          2.        ]
 [ 3.23606798 -1.23606798]]

A^10 =  [[34. 55.]
 [55. 89.]]


This shows that diagonalization can be effectively used to quickly compute powers of a matrix.